test/src/test/java/org/corfudb/integration/StreamAddressDiscoveryIT.java

```diff
@@ -1002,4 +1002,78 @@ public void testStreamRebuiltWithHoleAsFirstEntryAfterTrimNoCP() throws Exceptio
             shutdownCorfuServer(server);
         }
     }
+
+    /**
+     * Test rebuilding a stream from a new runtime, whenever the last address before the checkpoint
+     * was a hole.
+     *
+     * This case is interesting to test as addresses that become holes are discarded by the streamView,
+     * while sequencer's are agnostic of this info, hence, take it into account for trim mark computation.
+     */
+    @Test
+    public void testCheckpointWhenLastAddressIsHole() throws Exception {
+        final int numEntries = 10;
+
+        // Run Corfu Server
+        Process server = runDefaultServer();
+
+        // Create Runtime
+        runtime = createDefaultRuntime();
+
+        // Instantiate streamA and streamB as maps
+        final String streamA = "streamA";
+        final String streamB = "streamB";
+        Map<String, Integer> mA = createMap(runtime, streamA);
+        Map<String, Integer> mB = createMap(runtime, streamB);
+
+        // Write 10 Entries to streamA
+        for (int i = 0; i < numEntries; i++) {
+            mA.put(String.valueOf(i), i);
+        }
+
+        // Force a hole as the last update to streamA
+        Token token = runtime.getSequencerView().next(CorfuRuntime.getStreamID(streamA)).getToken();
+        LogData hole = LogData.getHole(token);
+        runtime.getLayoutView().getRuntimeLayout()
+                .getLogUnitClient("tcp://localhost:9000")
+                .write(hole);
+
+        // Write 10 Entries to streamB
+        for (int i = 0; i < numEntries; i++) {
+            mB.put(String.valueOf(i), i);
+        }
+
+        // Start a CheckpointWriter for streamA
+        CheckpointWriter cpwA = new CheckpointWriter(runtime, CorfuRuntime.getStreamID(streamA),
+                "checkpointer-Test", mA);
+        Token cpTokenA = cpwA.appendCheckpoint();
+
+        // Start a CheckpointWriter for streamB
+        CheckpointWriter cpwB = new CheckpointWriter(runtime, CorfuRuntime.getStreamID(streamB),
+                "checkpointer-Test", mB);
+        cpwB.appendCheckpoint();
+
+        // Add an update to streamA after checkpoint
+        mA.put(String.valueOf(numEntries), numEntries);
+
+        // Trim the log at B's CPToken
+        runtime.getAddressSpaceView().prefixTrim(cpTokenA);
+
```

Comment:
Restart, or flush the server caches after the trim

Reply:
Done.